pkg/webhook/cloudevents.go

```diff
@@ -0,0 +1,279 @@
+		return fmt.Errorf("invalid content type: %s", contentType)
+	}
+
+	// If secret is configured, validate it
+	// CloudEvents doesn't have a standard authentication mechanism,
+	// so we support a simple query parameter or header-based secret
+	if c.secret != "" {
+		// Check for secret in query parameter
+		secret := r.URL.Query().Get("secret")
+		if secret == "" {
+			// Check for secret in header
+			secret = r.Header.Get("X-Webhook-Secret")
+		}
+
+		if secret == "" {
+			return fmt.Errorf("missing webhook secret")
+		}
+
+		if subtle.ConstantTimeCompare([]byte(secret), []byte(c.secret)) != 1 {
+			return fmt.Errorf("invalid webhook secret")
+		}
+	}
+
+	return nil
+}
+
+// Parse processes the CloudEvents webhook payload and returns a WebhookEvent
+func (c *CloudEventsWebhook) Parse(r *http.Request) (*argocd.WebhookEvent, error) {
+	webhookLogger := log.Log().WithFields(logrus.Fields{
+		"logger": "cloudevents-webhook",
+	})
+	ctx := log.ContextWithLogger(r.Context(), webhookLogger)
+	logCtx := log.LoggerFromContext(ctx)
+
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read request body: %w", err)
+	}
+
+	// CloudEvents specification: https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md
+	var payload struct {
+		SpecVersion     string                 `json:"specversion"`
+		Type            string                 `json:"type"`
+		Source          string                 `json:"source"`
+		Subject         string                 `json:"subject"`
+		ID              string                 `json:"id"`
+		Time            string                 `json:"time"`
+		DataContentType string                 `json:"datacontenttype"`
+		Data            map[string]interface{} `json:"data"`
+	}
+
+	if err := json.Unmarshal(body, &payload); err != nil {
+		return nil, fmt.Errorf("failed to parse webhook payload: %w", err)
+	}
+
+	logCtx.Tracef("Parsed CloudEvents: specversion=%s, type=%s, source=%s, subject=%s",
+		payload.SpecVersion, payload.Type, payload.Source, payload.Subject)
+
+	// Validate CloudEvents spec version
+	if payload.SpecVersion == "" {
+		return nil, fmt.Errorf("missing CloudEvents specversion")
+	}
+
+	if payload.Type == "" {
+		return nil, fmt.Errorf("missing CloudEvents type")
+	}
+
+	// Parse the event based on the type
+	return c.parseEvent(&payload, ctx)
+}
+
+// parseEvent extracts webhook event data from CloudEvents payload
+func (c *CloudEventsWebhook) parseEvent(payload *struct {
+	SpecVersion     string                 `json:"specversion"`
+	Type            string                 `json:"type"`
+	Source          string                 `json:"source"`
+	Subject         string                 `json:"subject"`
+	ID              string                 `json:"id"`
+	Time            string                 `json:"time"`
+	DataContentType string                 `json:"datacontenttype"`
+	Data            map[string]interface{} `json:"data"`
```

Comment:
can we extract the payload struct into a named struct, to avoid duplication in `parse()` and `parseEvent()`?